subsys/mgmt: Add UDP transport for SMP

aunsbjerg opened the pull request:
This makes it possible to perform `mcumgr` operation like DFU over ethernet instead of bluetooth or serial. The `mcumgr` tool already has support for UDP transport.

Commit:
mgmt: smp: add UDP transport for SMP

Adds a UDP driver dedicated to transporting mcumgr SMP requests and
responses.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

carlescufi commented:
@aunsbjerg thanks for the PR!
I will let @de-nordic and @nvlsianpu decide over this, but would it make sense to target the upstream mcumgr repository for (parts of) this PR?

aunsbjerg commented:
@carlescufi I don't think it's necessary to add anything to upstream mcumgr, as we're basically just adding a transport that's already supported by the mcumgr-cli. I might be wrong though.

de-nordic commented:
> 
> 
> @aunsbjerg thanks for the PR!
> I will let @de-nordic and @nvlsianpu decide over this, but would it make sense to target the upstream mcumgr repository for (parts of) this PR?

The PR only adds transport, similarly to smp_uart, smp_bt sources. The management part is still performed by mcumgr lib. I think that this is the the right place for the code.

nvlsianpu commented:
I have been looking at other transport solution as well. This PR is in right place.

nvlsianpu commented:
@aunsbjerg Can you become one of code-owner of files you added or modified?

jukkar commented on samples/subsys/mgmt/mcumgr/smp_svr/prj_udp.conf:
```diff
@@ -0,0 +1,40 @@
+# Enable mcumgr.
```
I think you should make the UDP support in overlay conf file (overlay-udp.conf). Then it is much easier to enable it during compilation.

aunsbjerg replied:
I agree, but I think this would require a rewrite of the entire sample, as there currently is no base config on which to overlay additional settings. I can do a rewrite, but I can't really verify that the bluetooth sample still works as I don't have the nrf board.

jukkar replied:
> I agree, but I think this would require a rewrite of the entire sample, as there currently is no base config on which to overlay additional settings.

Yes, you would probably need to rethink the content of the conf files. You should place common config into prj.conf file and then just things related to features etc into overlay config file. I think this should not be a big effort and would make the sample more usable.

jukkar commented on samples/subsys/mgmt/mcumgr/smp_svr/src/main.c:
```diff
@@ -163,6 +168,24 @@ void main(void)
+
+	struct net_if *iface = net_if_get_default();
+
+	while (!net_if_is_up(iface)) {
```
This is a big no-no. We have infrastructure in zephyr to wait connectivity event and you should use that. See `samples/net/sockets/echo_client/src/echo-client.c` for details.

jukkar commented on subsys/mgmt/Kconfig:
```diff
@@ -57,18 +56,60 @@ config MCUMGR_SMP_UART_MTU
+	bool "UDP mcumgr SMP transport"
+	select MCUMGR
+	select NETWORKING
+	select NET_IPV4
```
Why not IPv6 support? From application point of view having IPv6 support is trivial. We are trying to make the IPv6 as a 1st class citizen and this is not helping.

jukkar replied:
Instead of selecting IPv4 here, you could leave this empty and let the application writer to select what protocol he/she wants to use.

jukkar commented on subsys/mgmt/Kconfig:
```diff
@@ -57,18 +56,60 @@ config MCUMGR_SMP_UART_MTU
+	  Scheduling priority of the SMP UDP listening thread.
+
+config MCUMGR_SMP_UDP_MTU
+	int "UDP SMP MTU"
```
This does not seem to be a network MTU but just some max buffer size when receiving data. Please adjust the name and remove the MTU as it is quite confusing to read.

aunsbjerg replied:
The term MTU is generally used in the mgmt SMP subsys, so the naming I chose is based on existing values like `MCUMGR_SMP_SHELL_MTU`. Do you still want me to change the naming?

jukkar replied:
It the term is used generally in the SMP, then that is fine.

jukkar commented on samples/subsys/mgmt/mcumgr/smp_svr/prj_udp.conf:
```diff
@@ -0,0 +1,40 @@
+# Network settings
+CONFIG_NETWORKING=y
+CONFIG_NET_MGMT=y
+CONFIG_NET_L2_ETHERNET_MGMT=y
```
Is this UDP transport only working with Ethernet?
If yes, then the max Ethernet MTU is 1500 bytes so you would not be able to receive 2027 byte UDP packets.

aunsbjerg replied:
I have changed the MTU to 1500 bytes. The reason for the ~2000 bytes is that the MTU in the mcumgr cli is hardcoded to 2000 bytes for UDP transport. I have not seen any packets that are more than ~600 bytes when uploading images, so 1500 should be okay.

jukkar commented on samples/subsys/mgmt/mcumgr/smp_svr/prj_udp.conf:
```diff
@@ -0,0 +1,40 @@
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_AUTO_INIT=y
+CONFIG_NET_CONFIG_NEED_IPV4=y
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="169.254.0.100"
```
Why such hard coded IP address? For IPv4, it is also possible to enable CONFIG_NET_IPV4_AUTO which enables IPv4 auto address configuration (RFC 3927)

aunsbjerg replied:
Honestly, I haven't been able to make CONFIG_NET_IPV4_AUTO work properly for me, so I opted for hardcoded addresses instead. I'm using the same adresses as the echo_client socket example.

jukkar replied:
> I haven't been able to make CONFIG_NET_IPV4_AUTO work properly for me,

Ok, it would be nice to know why IPv4 auto addresses are not working. Could you create an issue for this one so we can try to figure out if there is a real bug in it or if it is just a configuration issue.

aunsbjerg replied:
I just tested again and it does in fact work. I had some other issues earlier that I attributed to CONFIG_NET_IPV4_AUTO.

But, I'm still not convinced using automatic ip configuration is a good fit for this sample. The mcumgr tool needs to know the ip address of the device, so the sample needs to print that address and the user needs to copy that every time. And what about ipv6?

So, I suggest we stick with static addresses for this sample.

jukkar replied:
Sure, using static addresses is ok. I was just wondering why you have the address selected from 169.254/16 subnet which is allocated for IPv4 auto addresses.

dbkinder reviewed (approved):
+1 for doc change :)

de-nordic commented on subsys/mgmt/smp_udp.c:
```diff
@@ -0,0 +1,145 @@
+			printk("smp: recvfrom error: %i", errno);
+		}
+
+		if (k_sem_take(&smp_udp_abort_sem, K_NO_WAIT) == 0) {
```
Wouldn't volatile global lifetime variable suffice here instead of semaphore? What is benefit of using semaphore here?

de-nordic commented on subsys/mgmt/smp_udp.c:
```diff
@@ -0,0 +1,145 @@
+
+			zephyr_smp_rx_req(&smp_udp_transport, nb);
+		} else if (len < 0) {
+			printk("smp: recvfrom error: %i", errno);
```
missing newline?

aunsbjerg commented:
Thanks for the feedback. I will try and get the changes made during the weekend. I will also update the documentation for the sample and for the subsys in general where applicable.

carlescufi commented:
> Thanks for the feedback. I will try and get the changes made during the weekend. I will also update the documentation for the sample and for the subsys in general where applicable.

Thanks @aunsbjerg  for the PR and the feedback. I think you are right and nothing is required upstream, but I'd like @nvlsianpu and @de-nordic to approve.

Commit:
samples: subsys: mgmt: smp_svr: refactor sample

To prepare this sample supporting future SMP transports, the sample
code is now split into a main and bluetooth file. A common config
has been identified and application specific config is put into
overlay config files.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
samples: subsys: mgmt: smp_svr: add udp sample

This sample now supports SMP UDP transport.
Two config overlays have been added for ipv4 and ipv6, respectively.

The sample documentation has been completely revamped to be less
bluetooth focused and more general.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

aunsbjerg commented:
Alright, finally got around to actually reworking this according to feedback.

Sample application is refactored. Common configuration stuff is put into prj.conf and transport specific is put into overlays. Code is separated into main, bt and udp specific stuff. I do not have any nordic boards so I have not been able to test that the bluetooth samples still work. Can someone verify this for me?

I have added myself as codeowner for the smp_udp.c in subsys and for the sample application. It would be nice to have a co-owner for the sample for the bluetooth stuff.

chrta commented on samples/subsys/mgmt/mcumgr/smp_svr/overlay-udp-ipv6.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_NET_BUF_USER_DATA_SIZE=24
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::1"
+CONFIG_ETH_MCUX_0_UNIQUE_MAC=y
```
This settings seems to be the only one that is specific for the k64f board. Is it required to run on this board? If not, can you remove it here and in the ipv4 overlay file?

jukkar commented on subsys/mgmt/smp_udp.c:
```diff
@@ -0,0 +1,175 @@
+	struct sockaddr *src_ud = net_buf_user_data(src);
+	struct sockaddr *dst_ud = net_buf_user_data(dst);
+
+	*dst_ud = *src_ud;
```
Better to use `net_ipaddr_copy(dst_ud, src_ud);` here as it handles the unaligned memory access (not really an issue here but good to have anyway).

jukkar commented on subsys/mgmt/smp_udp.c:
```diff
@@ -0,0 +1,175 @@
+{
+	struct sockaddr *addr;
+
+#if defined(CONFIG_MCUMGR_SMP_UDP_IPV6)
```
This of course works but have you considered allowing both IPv4 and IPv6 support at the same time? In other parts of Zephyr there are similar patterns where one could have either
* IPv4 only
* IPv6 only
* dual IPv4 and IPv6

support. Now the code only supports either IPv4 or IPv6 and you must know in advance how the user will access the system.

aunsbjerg replied:
I've been thinking a bit about this. I agree it would be nice to have both transports supported at the same time. I see a few ways of going about this:
* Single smp transport and single thread. Use select() for polling both sockets.
* Single smp transport and one thread per network transport.
* One smp transport for each network transport.

The single smp transport approaches will work fine assuming the user uses one network transport at any given moment. Otherwise, we theoretically risk mixing smp fragments from the two network transports.

Having one smp transport per network transport eliminates the risk of mixing smp fragments at the cost of some extra code.

I'm leaning towards a single smp transport to keeps things simple. Any thoughts on this?

jukkar replied:
> leaning towards a single smp transport to keeps things simple

That sounds just fine

aunsbjerg replied:
Ended up adding separate smp transports for ipv4 and ipv6 as is was very little extra work.

jukkar commented on subsys/mgmt/smp_udp.c:
```diff
@@ -0,0 +1,175 @@
+			ud = net_buf_user_data(nb);
+
+			if (IS_ENABLED(CONFIG_MCUMGR_SMP_UDP_IPV6)) {
+				*ud = *(struct sockaddr *)net_sin6(&addr);
```
This could lead to unaligned memory access, use `net_ipaddr_copy()` instead.

jukkar commented on subsys/mgmt/smp_udp.c:
```diff
@@ -0,0 +1,175 @@
+{
+	stop_smp_udp = true;
+	close(smp_udp_sock);
+
```
Should we terminate the receive thread in close?

aunsbjerg commented:
Updated commits according to feedback from @jukkar 
Notably, both ipv4 and ipv6 support can be enabled at the same time or independently of each other. 
UDP ipv4 and ipv6 samples have been merged into a single sample.

jukkar commented on subsys/mgmt/Kconfig:
```diff
@@ -64,16 +64,77 @@ config MCUMGR_SMP_UART_MTU
+	  Enable SMP UDP using IPv4 addressing.
+	  Can be enabled alongside IPv6 addressing.
+
+config MCUMGR_SMP_UDP_IPV6
```
Do you need this extra level of configuration? What I mean is that now it is possible to have the device to have IPv6 enabled, but SMP UDP to not support it. Perhaps this is ok, it really depends on the use cases.
For example in networking libs and apps, it is usually enough to just enable IPv6 which then enables IPv6 usage in corresponding lib/app. So usually it is not needed to turn off IPv6 in app if the system globally supports IPv6.

aunsbjerg replied:
It's not necessary, but I figured one might want to disable ipv4/ipv6 for SMP independently of the network stack configuration. I can remove these config items and use the network config values instead if that's preferable.

jukkar replied:
This is more or less up to you and the use cases. If you have a good use case for it, that is fine. It was just not obvious by looking the patch so wanted to clarify it.

jukkar reviewed (approved):
LGTM

nvlsianpu commented on samples/subsys/mgmt/mcumgr/smp_svr/overlay-bt.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_FILE_SYSTEM_LITTLEFS=y
+
+# Enable file system commands
+CONFIG_MCUMGR_CMD_FS_MGMT=n
```
CONFIG_MCUMGR_CMD_FS_MGMT=y

aunsbjerg replied:
See #24444
